storage/engine: tweak RocksDB options

petermattis opened the pull request:
Together these tweaks reduce the time for `block_writer` to insert 1m
rows from 250s to 225s (-10%).

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7532)

<!-- Reviewable:end -->

petermattis committed:
storage/engine: tweak RocksDB options

Enabling `level_compaction_dynamic_level_bytes` seems to be the big win.

petermattis committed:
server: bump defaultMemtableBudget to 32 MB

petermattis commented:
These are pulled from my `pmattis/rocksdb-tuning` branch.

petermattis committed:
storage/engine: enable max_subcompactions

Set max_subcompations to `num_cpu/2`.

tschottdorf commented on storage/engine/rocksdb/db.cc:
```diff
@@ -1353,15 +1353,9 @@ DBStatus DBOpen(DBEngine **db, DBSlice dir, DBOptions db_opts) {
-
-  rocksdb::Options options(rocksdb::DBOptions(), cf_options);
+  rocksdb::Options options(rocksdb::GetOptions(db_opts.memtable_budget));
+  options.IncreaseParallelism(db_opts.num_cpu);
```
Any commentary here?

petermattis replied:
Which line? The `IncreaseParallelism` call allows multiple background compactions to be performed concurrently and ties into the `max_subcompactions` setting below which will use multiple threads to speed up a single compaction.

tschottdorf replied:
Mostly the second but generally all of them - I need a comment to explain why `(num_cpu / 2, 1)` was chosen and I think everyone else who's not in the RocksDB internals will appreciate it too. Super useful to know what your reasoning is when you turn random knobs.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/engine/rocksdb/db.cc, line 1368 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-KLSqCFCGnnwv4n9bMar:-KLSqCFDJJMIoi430i7p:1589016654) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc97dc47e56b2ede862902cf49ef267bdafea0f/storage/engine/rocksdb/db.cc#L1368)):*
> ```cpp
>   options.table_factory.reset(rocksdb::NewBlockBasedTableFactory(table_options));
> 
>   // File size settings: TODO(marc): investigate and determine better long-term settings:
> ```

this TODO is sitting on the wrong thing now

---

*[storage/engine/rocksdb/db.cc, line 1358 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-KLSqL89BJaqar6c626w:-KLSqL89BJaqar6c626x:-1709433982) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2868189cbfa6813924461da6acf08f5c4b9f90c/storage/engine/rocksdb/db.cc#L1358)):*
> ```cpp
>   rocksdb::Options options(rocksdb::GetOptions(db_opts.memtable_budget));
>   options.IncreaseParallelism(db_opts.num_cpu);
>   options.max_subcompactions = std::max<int>(db_opts.num_cpu / 2, 1);
> ```

science dog is surprised you needed to specify the type parameter

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-:-KLSqQvWT6NUj98qRigg:1417088721)*
<!-- Sent from Reviewable.io -->

petermattis committed:
storage/engine: tune max_bytes_for_level_base

Per RocksDB tuning guidelines, set `max_bytes_for_level_base` to the
estimated size of L0.

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/engine/rocksdb/db.cc, line 1368 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-KLSqCFCGnnwv4n9bMar:-KLSwMwv8A3JWnD8pvQn:1267029685) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc97dc47e56b2ede862902cf49ef267bdafea0f/storage/engine/rocksdb/db.cc#L1368)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this TODO is sitting on the wrong thing now
</details>
Removed as I've now done some tuning here.

---

*[storage/engine/rocksdb/db.cc, line 1357 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-KLScr6ijb860OtyGZ7Z-r3-1357:-KLSvkmbrIveaCAlxt44:1626901177) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2868189cbfa6813924461da6acf08f5c4b9f90c/storage/engine/rocksdb/db.cc#L1357)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Mostly the second but generally all of them - I need a comment to explain why `(num_cpu / 2, 1)` was chosen and I think everyone else who's not in the RocksDB internals will appreciate it too. Super useful to know what your reasoning is when you turn random knobs.
</details>
I added some commentary in various places.

---

*[storage/engine/rocksdb/db.cc, line 1358 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-KLSqL89BJaqar6c626w:-KLSvpEfI_AZkISh3yTU:1782943938) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2868189cbfa6813924461da6acf08f5c4b9f90c/storage/engine/rocksdb/db.cc#L1358)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> science dog is surprised you needed to specify the type parameter
</details>
It wasn't. Removed.

---

*[storage/engine/rocksdb/db.cc, line 1388 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-KLSvs4qQZsylhCPsfx8:-KLSvs4qQZsylhCPsfx9:-469213201) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b8c30935fccd39434ccf4f7ee0976aed6f9f744/storage/engine/rocksdb/db.cc#L1388)):*
Note that I added tuning of this field which was massively misconfigured. With the current settings we'll compute the value here as 32 MB while we previously had it configured as 512 MB.

> We usually recommend making L0->L1 as fast as possible by making the size of level 0 similar to size of level 1.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-:-KLT9amhbSNQfkbjoWwm:1417088721)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Unfortunately, this doesn't improve performance for the other K/V slowdown (#7280), but these seem like good changes nonetheless based on RocksDB tuning guidelines.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7532#-:-KLTyjHmRBJNHiAXblYO:928237250)*
<!-- Sent from Reviewable.io -->
